### PR TITLE
RFC: Adds speculative API specification

### PR DESCRIPTION
**Pull request by despeset**
Wildly speculative API for a micro-framework -- not to be merged ( at least, until the API is fully implemented :bowtie: )

Discuss.

### CODE REVIEW

**Comment by zachallaun** [comment]
I'M FOR IT.

File: example-app.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+post(app, "/") do req, res
+    # ...
+end
+
+# Route with URL vars
+route(app, "/hello/<id::Int>/<name::%[a-z]{3}[0-9]{2}>") do req, res
+    req.resources[:id] == 99 ? string("Go to hell ", req.resources[:name]) : string("Hello ", req.resources[:name])
+end
+
+route(app, GET, "/") do req, res
```
Comment (despeset) [inline example-app.jl]: Does this `GET` stynax qualify as Type abuse?  Would it be better if it were a symbol?

Comment (StefanKarpinski) [inline example-app.jl]: I don't know what GET is, so hard to say. It could just be an integer. Or it could be a type object that's being dispatched on.

Comment (zachallaun) [inline example-app.jl]: My thought was that this was an integer.

Comment (chuckha) [inline example-app.jl]: What about [GET] or [POST] for consistency?

Comment (zachallaun) [inline example-app.jl]: I don't believe it's poor style in Julia to have multiple ways to call a function. Defining this form will probably be as simple as

``` jl
function route(handler::Function, app::Micro.App, method::Micro.Method, resource::String)
    route(handler, app, [method], resource)
end
```

Comment (StefanKarpinski) [inline example-app.jl]: If these methods are just integers, you could make them all powers of two and just OR them together instead: `GET | PUT`. Efficiency doesn't much matter for configuration, but it is actually a nice interface. The type could be changed later to something more complicated and still use the same syntax even.

Comment (zachallaun) [inline example-app.jl]: That's… awesome.

**Comment by zachallaun** [comment]
This branch has been rebased locally and merged into master.

**Pull request closed without merging**